### PR TITLE
[mlir][Vector] Add vector bitwidth target to xfer op flattening

### PR DESCRIPTION
**Pull request by dcaballe**
This PR adds an optional bitwidth parameter to the vector xfer op flattening transformation so that the flattening doesn't happen if the trailing dimension of the read/writen vector is larger than this bitwidth (i.e., we are already able to fill at least one vector register with that size).

### CODE REVIEW

File: mlir/test/lib/Dialect/Vector/TestVectorTransforms.cpp [hunk 1]
```diff
@@ -480,7 +480,8 @@ struct TestFlattenVectorTransferPatterns
   }
   void runOnOperation() override {
     RewritePatternSet patterns(&getContext());
-    populateFlattenVectorTransferPatterns(patterns);
+    constexpr unsigned targetVectorBitwidth = 512;
```
Comment (banach-space) [inline mlir/test/lib/Dialect/Vector/TestVectorTransforms.cpp:483]: Could you make this a pass option and the toggle it in the test to demonstrate how it impacts the pattern?

File: mlir/test/Dialect/Vector/vector-transfer-flatten.mlir [hunk 2]
```diff
@@ -1,4 +1,5 @@
 // RUN: mlir-opt %s -test-vector-transfer-flatten-patterns -split-input-file | FileCheck %s
+// RUN: mlir-opt %s -test-vector-transfer-flatten-patterns=target-vector-bitwidth=128 -split-input-file | FileCheck %s --check-prefix=CHECK-128B
```
Comment (banach-space) [inline mlir/test/Dialect/Vector/vector-transfer-flatten.mlir:2]:
```suggestion
// RUN: mlir-opt %s -test-vector-transfer-flatten-patterns -split-input-file | FileCheck %s --check-prefixex=CHECK,CHECK-DEFAULT

// RUN: mlir-opt %s -test-vector-transfer-flatten-patterns=target-vector-bitwidth=128 -split-input-file | FileCheck %s --check-prefix=CHECK,CHECK-128B
```

This way, you don't need to add `CHECK-128B` to every "sub-file" in this file and only update checks for e.g. `transfer_read_dims_mismatch_non_zero_indices`:
```
// CHECK-DEFAULT-LABEL: func @transfer_read_dims_match_contiguous
//       CHECK-DEFAULT:   memref.collapse_shape
// ...
// CHECK-128B-LABEL: func @transfer_read_dims_match_contiguous
//       CHECK-128B:   memref.collapse_shape
// ...
```
https://llvm.org/docs/CommandGuide/FileCheck.html#options

[nit] Side note. IMHO, Once "prefixes" are involved, "CHECK" in prefix name becomes noise.

File: mlir/test/Dialect/Vector/vector-transfer-flatten.mlir [hunk 3]
```diff
@@ -66,7 +76,7 @@ func.func @transfer_read_dims_mismatch_non_zero_indices(
                      %m_out: memref<1x2x6xi32>) {
   %c0 = arith.constant 0 : index
   %c0_i32 = arith.constant 0 : i32
-  %2 = vector.transfer_read %m_in[%c0, %idx_1, %idx_2, %c0], %c0_i32 {in_bounds = [true, true, true]} : 
+  %2 = vector.transfer_read %m_in[%c0, %idx_1, %idx_2, %c0], %c0_i32 {in_bounds = [true, true, true]} :
     memref<1x43x4x6xi32>, vector<1x2x6xi32>
   vector.transfer_write %2, %m_out[%c0, %c0, %c0] {in_bounds = [true, true, true]} :
     vector<1x2x6xi32>, memref<1x2x6xi32>
```
Comment (banach-space) [inline mlir/test/Dialect/Vector/vector-transfer-flatten.mlir:82]: Something for the future ...

This example, on a machine with vectors which are 128 bits wide (e.g. Arm) would actually benefit from flattening. With 6 elements, we'd use 1.5 vector registers. And and with 12 elements, we'd use 3. That would be better utilization.

Would that make sense as a TODO? (not asking for it in this patch)

**Review by banach-space — APPROVED** [review]
Thanks, makes sense! LGTM

File: mlir/lib/Dialect/Vector/Transforms/VectorTransferOpTransforms.cpp [hunk 4]
```diff
@@ -554,6 +561,12 @@ class FlattenContiguousRowMajorTransferReadPattern
     // If this is already 0D/1D, there's nothing to do.
     if (vectorType.getRank() <= 1)
       return failure();
+    if (!vectorType.getElementType().isSignlessIntOrFloat())
+      return failure();
+    unsigned trailingVectorDimBitwidth =
+        vectorType.getShape().back() * vectorType.getElementTypeBitWidth();
+    if (trailingVectorDimBitwidth >= targetVectorBitwidth)
+      return failure();
     if (!vector::isContiguousSlice(sourceType, vectorType))
```
Comment (hanhanW) [inline mlir/lib/Dialect/Vector/Transforms/VectorTransferOpTransforms.cpp:570]: There could be cases that only partial slice is contiguous. In this case, we could flatten trailing dims. I wonder if we will relax this a little more in the near future?

File: mlir/lib/Dialect/Vector/Transforms/VectorTransferOpTransforms.cpp [hunk 5]
```diff
@@ -702,6 +731,11 @@ class FlattenContiguousRowMajorTransferWritePattern
     rewriter.eraseOp(transferWriteOp);
     return success();
   }
+
+private:
+  // Minimum bitwidth that the trailing vector dimension should have after
+  // flattening.
+  unsigned targetVectorBitwidth;
```
Comment (hanhanW) [inline mlir/lib/Dialect/Vector/Transforms/VectorTransferOpTransforms.cpp:738]: `after flattening`? It seems not correct with the implementation. With `targetVectorBitWidth=128` and `vector<1x2x6xi32>` type, it becomes `vector<12xi32>` after flattening; the bit-width of trailing dim is `384`. It sounds like we should flatten it with the configuration.

How about update it to `Maximum bitwidth that ... before flattening`?

**Review by hanhanW — APPROVED** [review]

File: mlir/test/Dialect/Vector/vector-transfer-flatten.mlir (hunk 3 above)
Comment (dcaballe) [reply mlir/test/Dialect/Vector/vector-transfer-flatten.mlir:82]: Yes, that's the ultimate goal: partial and more targeted flattening but one step at a time. We first have to flatten the producer/consumers of these xfer ops to make sure we don't generate ops to reshape the vector.

File: mlir/lib/Dialect/Vector/Transforms/VectorTransferOpTransforms.cpp (hunk 4 above)
Comment (dcaballe) [reply mlir/lib/Dialect/Vector/Transforms/VectorTransferOpTransforms.cpp:570]: Yes, that's the ultimate goal: partial and more targeted flattening but one step at a time. We first have to flatten the producer/consumers of these xfer ops to make sure we don't generate ops to reshape the vector.

File: mlir/test/Dialect/Vector/vector-transfer-flatten.mlir (hunk 2 above)
Comment (dcaballe) [reply mlir/test/Dialect/Vector/vector-transfer-flatten.mlir:2]: I'm not sure I follow. I added the check for wider coverage. We can leave tests without check rules for a particular run. The only difference I see is that we could have a single label check that is checked in both runs but that doesn't seem to be what you meant

Comment (banach-space) [reply mlir/test/Dialect/Vector/vector-transfer-flatten.mlir:2]: I was referring to the CHECK lines that are duplicated, eg for “ @transfer_read_0d”. But now I see that the duplication is only happening for negative tests, right? Sorry, was reading this in a rush and missed the finer details :(

**Pull request merged**